lxd/storage/drivers/interface.go

```diff
@@ -0,0 +1,65 @@
+package drivers
+
+import (
+	"github.com/lxc/lxd/lxd/migration"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/shared/api"
+)
+
+// VolumeType represents a storage volume type.
+type VolumeType string
+
+// VolumeTypeImage represents an image storage volume.
+const VolumeTypeImage = VolumeType("images")
+
+// VolumeTypeCustom represents a custom storage volume.
+const VolumeTypeCustom = VolumeType("custom")
+
+// VolumeTypeContainer represents a container storage volume.
+const VolumeTypeContainer = VolumeType("containers")
+
+// VolumeTypeVM represents a virtual-machine storage volume.
+const VolumeTypeVM = VolumeType("virtual-machines")
+
+// ContentType indicates the format of the volume.
+type ContentType string
+
+// ContentTypeFS indicates the volume will be populated with a mountabble filesystem.
+const ContentTypeFS = ContentType("fs")
+
+// ContentTypeBlock indicates the volume will be a block device and its contents and we do not
+// know which filesystem(s) (if any) are in use.
+const ContentTypeBlock = ContentType("block")
+
+// driver is the extended internal interface.
+type driver interface {
+	Driver
+
+	init(name string, config map[string]string, volIDFunc func(volType VolumeType, volName string) (int64, error))
+}
+
+// Driver represents a low-level storage driver.
+type Driver interface {
+	// Internal.
+	Name() string
+	Version() string
+	OptimizedImages() bool
+	HasVolume(volType VolumeType, volName string) bool
+
+	// Pool.
+	Create() error
+	Delete(op *operations.Operation) error
+	Mount() (bool, error)
+	Unmount() (bool, error)
+	GetResources() (*api.ResourcesStoragePool, error)
+
+	// Volumes.
+	CreateVolume(volType VolumeType, contentType ContentType, volName string, filler func(path string) error, op *operations.Operation) error
+	CreateVolumeFromCopy(volType VolumeType, volName string, srcVolType VolumeType, srcVolName string, op *operations.Operation) error
```

Comment:
Both `CreateVolume` and `CreateVolumeFromCopy` should actually be passed the Config map from api.StorageVolumesPost as there are driver specific information in there.

I just thought of it now thinking of how we'll be creating a "fs" contentType on LVM/CEPH where this involves the driver creating a block device of a configured size and creating the filesystem on it. Both size and filesystem are configurable at the pool level OR at the volume level, so the level config must be made available to the Create functions.

Reply:
Add `volConfig map[string]string` to those two functions and passed vol.Config through from lxdBackend.